beacon-chain/p2p/peers/scorers/peer_status.go

```diff
@@ -0,0 +1,143 @@
+package scorers
+
+import (
+	"errors"
+	"math"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/prysmaticlabs/prysm/beacon-chain/p2p/peers/peerdata"
+	p2ptypes "github.com/prysmaticlabs/prysm/beacon-chain/p2p/types"
+	pb "github.com/prysmaticlabs/prysm/proto/beacon/p2p/v1"
+	"github.com/prysmaticlabs/prysm/shared/timeutils"
+)
+
+var _ Scorer = (*PeerStatusScorer)(nil)
+
+// PeerStatusScorer represents scorer that evaluates peers based on their statuses.
+// Peer statuses are updated by regularly polling peers (see sync/rpc_status.go).
+type PeerStatusScorer struct {
+	config              *PeerStatusScorerConfig
+	store               *peerdata.Store
+	ourHeadSlot         uint64
+	highestPeerHeadSlot uint64
+}
+
+// PeerStatusScorerConfig holds configuration parameters for peer status scoring service.
+type PeerStatusScorerConfig struct{}
+
+// newPeerStatusScorer creates new peer status scoring service.
+func newPeerStatusScorer(store *peerdata.Store, config *PeerStatusScorerConfig) *PeerStatusScorer {
+	if config == nil {
+		config = &PeerStatusScorerConfig{}
+	}
+	return &PeerStatusScorer{
+		config: config,
+		store:  store,
+	}
+}
+
+// Score returns calculated peer score.
+func (s *PeerStatusScorer) Score(pid peer.ID) float64 {
+	s.store.RLock()
+	defer s.store.RUnlock()
+	return s.score(pid)
+}
+
+// score is a lock-free version of Score.
+func (s *PeerStatusScorer) score(pid peer.ID) float64 {
+	if s.isBadPeer(pid) {
+		return BadPeerScore
+	}
+	score := float64(0)
+	peerData, ok := s.store.PeerData(pid)
+	if !ok || peerData.ChainState == nil {
+		return score
+	}
+	if peerData.ChainState.HeadSlot < s.ourHeadSlot {
+		return score
+	}
+	// Calculate score as a ratio to the known maximum head slot.
+	// The closer the current peer's head slot to the maximum, the higher is the calculated score.
+	if s.highestPeerHeadSlot > 0 {
+		score = float64(peerData.ChainState.HeadSlot) / float64(s.highestPeerHeadSlot)
+		return math.Round(score*ScoreRoundingFactor) / ScoreRoundingFactor
+	}
+	return score
+}
+
+// IsBadPeer states if the peer is to be considered bad.
+func (s *PeerStatusScorer) IsBadPeer(pid peer.ID) bool {
+	s.store.RLock()
+	defer s.store.RUnlock()
+	return s.isBadPeer(pid)
+}
+
+// isBadPeer is lock-free version of IsBadPeer.
+func (s *PeerStatusScorer) isBadPeer(pid peer.ID) bool {
+	peerData, ok := s.store.PeerData(pid)
+	if !ok {
+		return false
+	}
+	// Mark peer as bad, if the latest error is one of the terminal ones.
+	terminalErrs := []error{
```

Comment:
maybe store this out of the function and in the general file. Rather than initializing it in the function

Reply:
Can I address this in the next PR (I will extend that field to include other errors there anyway)?